presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,268 @@
+import com.google.common.util.concurrent.SettableFuture;
+import io.prestosql.execution.StageInfo;
+import io.prestosql.execution.StageState;
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.execution.TaskManagerConfig;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.optimizations.PlanNodeSearcher;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNodeId;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.prestosql.execution.StageState.FINISHED;
+import static io.prestosql.execution.StageState.RUNNING;
+import static io.prestosql.execution.StageState.SCHEDULED;
+import static io.prestosql.execution.StageState.SCHEDULING_SPLITS;
+import static io.prestosql.spi.predicate.TupleDomain.columnWiseUnion;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, SettableFuture<TupleDomain<String>>> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Supplier<List<StageInfo>>> queries = new ConcurrentHashMap<>();
+    private final ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void registerQuery(QueryId queryId, Supplier<List<StageInfo>> stageInfoSupplier)
```

Comment:
nit: you could potentially pass just `SqlQueryExecution` here. `DynamicFilterService` can extract it's required info itself. This could simplify this method params

Reply:
The reason I had used `Supplier<List<StageInfo>>` as an argument instead of `SqlQueryExecution` is to simplify writing unit tests. Mocking `SqlQueryExecution` seems to require writing a lot of code. What I can do is make this method `VisibleForTesting` and add a  `registerQuery(SqlQueryExecution sqlQueryExecution)` method for use in `SqlQueryExecution`